pkg/proctree/tree/tree.go

```diff
@@ -0,0 +1,190 @@
+		table.SetAutoFormatHeaders(true)
+		table.SetHeaderAlignment(tablewriter.ALIGN_LEFT)
+		table.SetAlignment(tablewriter.ALIGN_LEFT)
+		table.SetHeaderLine(true)
+		table.SetBorder(true)
+		return table
+	}
+
+	table := newTable()
+
+	var lines [][]string
+
+	for processHash, process := range t.processes {
+		pid := fmt.Sprintf("%v", process.GetInfo().Pid)
+		parentPid := fmt.Sprintf("%v", t.GetParentPID(processHash))
+		cmd := process.GetCommand().Name
+		parentCmd := t.GetParent(processHash).GetCommand().Name
+		lines = append(lines, []string{parentPid, pid, parentCmd, cmd})
+	}
+
+	sort.Slice(lines, func(i, j int) bool {
+		firstOne, _ := strconv.Atoi(lines[i][0])
+		firstTwo, _ := strconv.Atoi(lines[j][0])
+		secondOne, _ := strconv.Atoi(lines[i][1])
+		secondTwo, _ := strconv.Atoi(lines[j][1])
+		if firstOne != firstTwo {
+			return firstOne < firstTwo
+		}
+		return secondOne < secondTwo
+	})
+
+	for _, line := range lines {
+		table.Append(line)
+	}
+	table.Render()
+
+	return buffer.String()
+}
+
+// NewTree returns a new tree
+func NewTree() *Tree {
+	return &Tree{
+		processes: make(map[uint32]*nodes.Process),
+		mutex:     &sync.RWMutex{},
+	}
+}
+
+// GetOrCreateProcess returns the process indexed by given hash, or creates a new one if not found.
+func (t *Tree) GetOrCreateProcess(hash uint32) *nodes.Process {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if _, ok := t.processes[hash]; !ok {
+		t.processes[hash] = nodes.NewProcess(hash)
+	}
+
+	return t.processes[hash]
+}
+
+// GetProcess returns the process indexed by given hash.
+func (t *Tree) GetProcess(hash uint32) (*nodes.Process, bool) {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	process, ok := t.processes[hash]
+	return process, ok
+}
+
+// GetParent returns the process indexed by given hash parent.
+func (t *Tree) GetParent(hash uint32) *nodes.Process {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	process, ok := t.processes[hash]
+	if !ok {
+		return nodes.NewProcess(0) // empty process node
+	}
+	parentHash := process.GetParentHash()
+	if parentHash == 0 {
+		return nodes.NewProcess(0) // empty process node
+	}
+	parent, ok := t.processes[parentHash]
+	if !ok {
+		return nodes.NewProcess(0) // empty process node
+	}
+
+	return parent
+}
+
+// GetParentPID returns the process indexed by given hash parent PID.
+func (t *Tree) GetParentPID(hash uint32) int32 {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	process, ok := t.processes[hash]
+	if !ok {
+		return 0
+	}
+	parentHash := process.GetParentHash()
+	if parentHash == 0 {
+		return 0
+	}
+	parent, ok := t.processes[parentHash]
+	if !ok {
+		return 0
+	}
+
+	return parent.GetInfo().Pid
+}
+
+// GetChildren returns the process indexed by given hash children.
+func (t *Tree) GetChildren(hash uint32) []*nodes.Process {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	process, ok := t.processes[hash]
+	if !ok {
+		return []*nodes.Process{} // empty slice
+	}
+
+	childrenHashes := process.GetChildrenHashes()
```

Comment:
I dont think so because of runtime memory usage and garbage collection needs. A hash is a hash, is not pointing to something that might or might not exist in memory (which could cause runtime issues), depending on the behavior of the system at execution time. 

Also, with this approach there is a single cache that can be maximized and does not need to be garbage collected, there are no pointers to be managed and updated over and over, no locking needs for the updates (besides adding the hash to the structure).

Either way, I never finished the code (up to a point a signature could use) and was more of seeking for feedback on the data structures, namings, things like that. I do feel the frustration of having the code re-done. No shame for me if my approach is worst, at least we will have given way more thought to it.

Also, please do feel free to merge anything else other than my code if you think my approach is not optimal.

Reply:
I see some significant pros for you approach.
As you said, this way the possibility of having a pointer to an object in the memory somewhere which prevents garbage collecting is much smaller.
This is also preventing the need for accessing multiple nodes when we want to clean one from the tree (for example upon reaching maximum memory we allow the tree to have).
The big draw-back here is the fact that this is not a tree - you can't iterate through it without the "tree" object. But this is also better, because it makes the responsibilities of the nodes much narrower, basically only data keeping objects, making them much easier to manage and update in the future.
Overall, I support the new approach of @rafaeldtinoco more then my previous one using the pointers.
Still open to hear more opinions though :)

Reply:
> The big draw-back here is the fact that this is not a tree - you can't iterate through it without the "tree" object. But this is also better, because it makes the responsibilities of the nodes much narrower, basically only data keeping objects, making them much easier to manage and update in the future.

THIS ^. Thanks for the support then, lets keep discussing! I'll rebase/change things based on our recent talks and we do another review round.